### PR TITLE
Error helpers + small refactors in `cmdlineTests.sh`

### PR DESCRIPTION
**Pull request by cameel**
A bunch of small tweaks to improve error reporting extracted out of my upcoming PR for #11877.

The biggest part is a helper Bash function that takes a command, and, if that command fails, prints stdout, stderr and exits. Doing this manually after every command adds a lot of boilerplate. I'll have 5 commands to handle so at this point it's more efficient to just write one reusable helper. It's also already useful in `cmdlineTests.sh` because there are a lot of places where we expect success and stderr gets sent to `/dev/null`. The helper will make debugging test failures there easier.

Other than that there are a few small refactors: `fail` helper, removal of `--force` from `rm`, unnecessary `set -e` in a few places and one missing `rm -r "$SOLTMPDIR"` (finally my `/tmp` won't be filled with garbage after running these tests). I have also enabled `set -o pipefail` - looks like it does not cause any spurious failures in this script.

### CODE REVIEW

File: test/cmdlineTests.sh [hunk 1]
```diff
@@ -471,52 +469,50 @@ SOLTMPDIR=$(mktemp -d)
...
 
 printTask "Testing long library names..."
-echo '' | "$SOLC" - --link --libraries aveeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeerylonglibraryname=0x90f20564390eAe531E810af625A22f51385Cd222 >/dev/null
+echo '' | msg_on_error --no-stdout "$SOLC" - --link --libraries aveeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeerylonglibraryname=0x90f20564390eAe531E810af625A22f51385Cd222
 
 printTask "Testing linking itself..."
 SOLTMPDIR=$(mktemp -d)
 (
     cd "$SOLTMPDIR"
-    set -e
```
Comment (chriseth) [inline test/cmdlineTests.sh:488]: Why can this be removed?

File: scripts/common.sh [hunk 2]
```diff
@@ -33,6 +33,79 @@ else
     function printLog() { echo "$(tput setaf 3)$1$(tput sgr0)"; }
 fi
 
+function fail() {
```
Comment (leonardoalt) [inline scripts/common.sh]:
```suggestion
function fail()
{
```

File: scripts/common.sh [hunk 3]
```diff
@@ -33,6 +33,79 @@ else
...
+    printError "$@"
+    return 1
```
Comment (chriseth) [inline scripts/common.sh:39]: is this essentially the same as `exit 1`?

Comment (leonardoalt) [inline scripts/common.sh:39]: Why return instead of exit?

File: test/cmdlineTests.sh (hunk 1 above)
Comment (cameel) [reply test/cmdlineTests.sh:488]: Because this enables exiting on errors and we already do that at the top of the file. I suspect this is just leftover after a `set +e`/`set -e` pair that might have been here in the past.

File: scripts/common.sh (hunk 3 above)
Comment (cameel) [reply scripts/common.sh:39]: Same but better :) With `exit 1` you immediately exit from the script. With `return 1` you just return with a non-zero status giving the calling script control over what happens next. If the script has `set -e` (which all of our scripts either do or should be doing), it will exit by default but you can also silence the error with `|| true` or `set +e` and do some error handling if you need to.

Basically, this makes the error work more like throwing an exception which bubbles up the stack than a hard exit.

Comment (cameel) [reply scripts/common.sh:39]: Answered in https://github.com/ethereum/solidity/pull/12076#discussion_r721323664 above.

File: test/cmdlineTests.sh (hunk 1 above)
Comment (chriseth) [reply test/cmdlineTests.sh:488]: My main question here is: Does the subshell `(...)` reset the flags or not?

Comment (cameel) [reply test/cmdlineTests.sh:488]: Entering the subshell does not reset them. The `set -e` at the beginning of the script will get inherited by the subshells. Also, any `set` invocations done inside subshell will not affect the rest of the script.

`set -e` in this subshell has no effect.

**Review by leonardoalt — APPROVED** [review]
